### PR TITLE
fix: stop loading entire file into memory to fix OOM errors

### LINKED ISSUE #2265: java.lang.OutOfMemoryError: Java heap space
**Describe the bug**
Hello, 

As we have been adding screenshot tests to more and more Android modules we wanted to parallelise our CI build to first build the test APK's.
After all modules have created their APK we run `./gradlew runFlank` 

We have about 14 modules thus-far containing screenshot tests.  
Averaging in size of approximately 60MB

Screenshot 2022-07-11 at 15 52 52

The build stops with a OutOfMemory.  
See full logs in `Build summary` section below.  

<details>
<summary>Build summary</summary>

>  > Task :configureFulladle
>  > Task :validateFladleConfig
>  > Task :writeConfigProps
>  > Task :execFlank
>  version: v21.11.0
>  revision: 1bf9b6a8ed1287141b66a75ec66b23813a02f2a8
>  session id: 08044f2b-9127-402c-9a6a-f5b6e7b51c13
>  AndroidArgs
>      gcloud:
>        results-bucket: test-lab-nxzauzcrb50h6-m6yq6i7d7dwni
>        results-dir: 2022-07-11_06-29-04.516092_hCGC
>        record-video: false
>        timeout: 15m
>        async: false
>        client-details:
>        network-profile: null
>        results-history-name: null
>        # Android gcloud
>        app: /xxx/xxxx/xxxxx/testlab/testlab-debug.apk
>        test: /xxx/xxxx/xxxxx/features/xxxx/build/outputs/apk/androidTest/debug/xxxx-debug-androidTest.apk
>        additional-apks:
>        auto-google-login: false
>        use-orchestrator: false
>        directories-to-pull:
>          - /storage/emulated/0/Download/screenshots/
>        grant-permissions: all
>        type: null
>        other-files:
>        scenario-numbers:
>        scenario-labels:
>        obb-files:
>        obb-names:
>        performance-metrics: false
>        num-uniform-shards: null
>        test-runner-class: null
>        test-targets:
>          - annotation com.xxxx.xxxxx.xxxxx.ShotTest
>        robo-directives:
>        robo-script: null
>        device:
>          - model: Pixel2
>            version: 29
>            locale: en
>            orientation: portrait
>        num-flaky-test-attempts: 0
>        test-targets-for-shard:
>        fail-fast: false
>        parameterized-tests: default
>      flank:
>        max-test-shards: 1
>        shard-time: -1
>        num-test-runs: 1
>        smart-flank-gcs-path: 
>        smart-flank-disable-upload: false
>        default-test-time: 120.0
>        use-average-test-time-for-new-tests: false
>        files-to-download:
>          - .*\.png$
>        test-targets-always-run:
>        disable-sharding: false
>        project: neba-android
>        local-result-dir: /xxx/xxxx/xxxxx/flank
>        full-junit-result: false
>        # Android Flank Yml
>        keep-file-path: false
>        additional-app-test-apks:
>          - test: /xxx/xxxx/xxxxx/features/xxxxx/build/outputs/apk/androidTest/debug/xxxxx-debug-androidTest.apk
>          - test: /xxx/xxxx/xxxxx/features/xxxxx/build/outputs/apk/androidTest/debug/xxxxx-debug-androidTest.apk
>          - test: /xxx/xxxx/xxxxx/features/xxxxx/build/outputs/apk/androidTest/debug/xxxxx-debug-androidTest.apk
>          - test: /xxx/xxxx/xxxxx/features/xxxxx/build/outputs/apk/androidTest/debug/xxxxx-debug-androidTest.apk
>          - test: /xxx/xxxx/xxxxx/features/xxxxx/build/outputs/apk/androidTest/debug/xxxxx-debug-androidTest.apk
>          - test: /xxx/xxxx/xxxxx/features/xxxxx/build/outputs/apk/androidTest/debug/xxxxx-debug-androidTest.apk
>          - test: /xxx/xxxx/xxxxx/features/xxxxx/build/outputs/apk/androidTest/debug/xxxxx-debug-androidTest.apk
>          - test: /xxx/xxxx/xxxxx/libraries/xxxxx/build/outputs/apk/androidTest/debug/xxxxx-debug-androidTest.apk
>          - test: /xxx/xxxx/xxxxx/libraries/xxxxx/build/outputs/apk/androidTest/debug/xxxxx-debug-androidTest.apk
>          - test: /xxx/xxxx/xxxxx/libraries/xxxxx/build/outputs/apk/androidTest/debug/xxxxx-debug-androidTest.apk
>          - test: /xxx/xxxx/xxxxx/libraries/xxxxx/build/outputs/apk/androidTest/debug/xxxxx-debug-androidTest.apk
>          - test: /xxx/xxxx/xxxxx/libraries/xxxxx/build/outputs/apk/androidTest/debug/xxxxx-debug-androidTest.apk
>            max-test-shards: 5
>        run-timeout: 30m
>        legacy-junit-result: false
>        ignore-failed-tests: false
>        output-style: verbose
>        disable-results-upload: false
>        default-class-test-time: 240.0
>        disable-usage-statistics: false
>        output-report: none
>        skip-config-validation: false
>        custom-sharding-json: 
>  RunTests
>   Smart Flank cache hit: 0% (0 / 23)
>    Shard times: 2760s
>   Smart Flank cache hit: 0% (0 / 1)
>    Shard times: 120s
>   Smart Flank cache hit: 0% (0 / 37)
>    Shard times: 4440s
>   Smart Flank cache hit: 0% (0 / 8)
>    Shard times: 960s
>   Smart Flank cache hit: 0% (0 / 3)
>    Shard times: 360s
>   Smart Flank cache hit: 0% (0 / 9)
>    Shard times: 1080s
>   Smart Flank cache hit: 0% (0 / 2)
>    Shard times: 240s
>   Smart Flank cache hit: 0% (0 / 4)
>    Shard times: 480s
>   Smart Flank cache hit: 0% (0 / 3)
>    Shard times: 360s
>   Smart Flank cache hit: 0% (0 / 1)
>    Shard times: 120s
>   Smart Flank cache hit: 0% (0 / 13)
>    Shard times: 1560s
>   Smart Flank cache hit: 0% (0 / 5)
>    Shard times: 600s
>   Smart Flank cache hit: 0% (0 / 211)
>    Shard times: 5040s, 5040s, 5040s, 5040s, 5160s
>    Saved 13 shards to /xxx/xxxx/xxxxx/flank/android_shards.json
>    Uploading [android_shards.json] to https://console.developers.google.com/storage/browser/test-lab-nxzauzcrb50h6-m6yq6i7d7dwni/2022-07-11_06-29-04.516092_hCGC/...
>    Uploading [testlab-debug.apk] to https://console.developers.google.com/storage/browser/test-lab-nxzauzcrb50h6-m6yq6i7d7dwni/2022-07-11_06-29-04.516092_hCGC/...
>    Uploading [xxxxx-debug-androidTest.apk] to https://console.developers.google.com/storage/browser/test-lab-nxzauzcrb50h6-m6yq6i7d7dwni/2022-07-11_06-29-04.516092_hCGC/....  Uploading [xxxxx-debug-androidTest.apk] to https://console.developers.google.com/storage/browser/test-lab-nxzauzcrb50h6-m6yq6i7d7dwni/2022-07-11_06-29-04.516092_hCGC/..  Uploading [xxxxx-debug-androidTest.apk] to https://console.developers.google.com/storage/browser/test-lab-nxzauzcrb50h6-m6yq6i7d7dwni/2022-07-11_06-29-04.516092_hCGC/..  Uploading [xxxxx-debug-androidTest.apk] to https://console.developers.google.com/storage/browser/test-lab-nxzauzcrb50h6-m6yq6i7d7dwni/2022-07-11_06-29-04.516092_hCGC/......  Uploading [xxxxx-debug-androidTest.apk] to https://console.developers.google.com/storage/browser/test-lab-nxzauzcrb50h6-m6yq6i7d7dwni/2022-07-11_06-29-04.516092_hCGC/...  Uploading [xxxxx-debug-androidTest.apk] to https://console.developers.google.com/storage/browser/test-lab-nxzauzcrb50h6-m6yq6i7d7dwni/2022-07-11_06-29-04.516092_hCGC/...........................
>    Uploading [xxxxx-debug-androidTest.apk] to https://console.developers.google.com/storage/browser/test-lab-nxzauzcrb50h6-m6yq6i7d7dwni/2022-07-11_06-29-04.516092_hCGC/...
>    Uploading [xxxxx-debug-androidTest.apk] to https://console.developers.google.com/storage/browser/test-lab-nxzauzcrb50h6-m6yq6i7d7dwni/2022-07-11_06-29-04.516092_hCGC/.....
>  java.lang.OutOfMemoryError: Java heap space
>	  at java.base/java.nio.file.Files.read(Files.java:3153)
>	  at java.base/java.nio.file.Files.readAllBytes(Files.java:3213)
>	  at ftl.util.FileReferenceKt.upload(FileReference.kt:33)
>	  at ftl.util.FileReferenceKt.uploadIfNeeded(FileReference.kt:28)
>	  at ftl.run.platform.android.UploadApksKt.upload(UploadApks.kt:35)
>	  at ftl.run.platform.android.UploadApksKt.upload(UploadApks.kt:27)
>	  at ftl.run.platform.android.UploadApksKt.access$upload(UploadApks.kt:1)
>	  at ftl.run.platform.android.UploadApksKt$upload$2$1$1.invokeSuspend(UploadApks.kt:23)
>	  at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
>	  at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
>	  at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
>	  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
>	  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
>	  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
>  Total run duration: 2m 40s
>	  - Preparation: 0m 6s
>  > Task :execFlank FAILED
>  FAILURE: Build failed with an exception.
>  * What went wrong:
>  Execution failed for task ':execFlank'.
>  > Process 'command '/usr/lib/jvm/java-11-openjdk-amd64/bin/java'' finished with non-zero exit value 15
>  * Try:
>  > Run with --stacktrace option to get the stack trace.
>  > Run with --info or --debug option to get more log output.
>  > Run with --scan to get full insights.
>  * Get more help at https://help.gradle.org
>  BUILD FAILED in 6m 27s
>  Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
>  You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
>  See https://docs.gradle.org/7.4.2/userguide/command_line_interface.html#sec:command_line_warnings
>  4 actionable tasks: 4 executed

</details>

```
java.lang.OutOfMemoryError: Java heap space
	at java.base/java.nio.file.Files.read(Files.java:3153)
	at java.base/java.nio.file.Files.readAllBytes(Files.java:3213)
	at ftl.util.FileReferenceKt.upload(FileReference.kt:33)
	at ftl.util.FileReferenceKt.uploadIfNeeded(FileReference.kt:28)
	at ftl.run.platform.android.UploadApksKt.upload(UploadApks.kt:35)
	at ftl.run.platform.android.UploadApksKt.upload(UploadApks.kt:27)
	at ftl.run.platform.android.UploadApksKt.access$upload(UploadApks.kt:1)
	at ftl.run.platform.android.UploadApksKt$upload$2$1$1.invokeSuspend(UploadApks.kt:23)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

Any idea on how to fix this?

**Additional context**
- Running on Gitlab shared runners
- Flank version: v21.11.0
- I saw possible related [ticket](https://github.com/Flank/flank/issues/1577) 

Thanks in advance.

### PR DESCRIPTION
**Pull request by joshafeinberg**
Fixes #2265

The code has been run internally in our company with no issues. We also have all passing tests
.

## Checklist

- [ ] Documented
- [ ] Unit tested
- [ ] Integration tests updated

### CODE REVIEW

**Comment by joshafeinberg** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by adamvduke** [comment]
Hi @joshafeinberg, sorry for the delay in looking at this. Since the creation of your PR, Flank has become a Google managed project, and requires signing the Google CLA.

Additionally, I've been trying your branch out locally, and the integration tests have a few failures. Are you able to run the `$ ./gradlew integrationTests` task successfully on your end?

File: test_runner/src/main/kotlin/ftl/client/google/GcStorage.kt [hunk 1]
```diff
@@ -114,6 +111,37 @@ object GcStorage {
...
     }
 
+    private fun File.uploadWithProgress(
+        bucket: String,
+        path: String,
+        name: String
+    ) {
+        // TODO #2136 handle messages with state
+        runWithProgress(
+            startMessage = "Uploading file [$name] to ${GCS_STORAGE_LINK + join(bucket, path).replace(name, "")}..",
```
Comment (adamvduke) [inline test_runner/src/main/kotlin/ftl/client/google/GcStorage.kt:121]: I looked into the integration test failures. The failure that I'm seeing looks like:

```
at app//utils.IntergrationTestsUtilsKt.assertContainsUploads(IntergrationTestsUtils.kt:60)
at app//integration.MultipleApksIT.flank full option run(MultipleApksIT.kt:45)
```

The definition of `assertContainsUploads`  looks like: https://github.com/Flank/flank/blob/1cec01d5c0e22caec0e85d5702c77154b1cef12e/integration_tests/src/test/kotlin/utils/IntergrationTestsUtils.kt#L59-L61

This new `startMessage` doesn't match the assertion. Removing the word 'file' and following space results in passing tests.

It's a brittle test, but I don't know the details of why it was written that way. My recommendation would be to change the value of `startMessage` to match the assertion, and at some point in the future, follow up on making the test less brittle.

**Comment by adamvduke** [comment]
@VictoriaAuto

File: test_runner/src/main/kotlin/ftl/client/google/GcStorage.kt (hunk 1 above)
Comment (joshafeinberg) [reply test_runner/src/main/kotlin/ftl/client/google/GcStorage.kt:121]: I tired updating the test but for some reason `./gradlew integrationTests` doesn't seem to run on my local machine. When I run the `MultipleApksIT` tests I pass so I believe the fix worked

Comment (adamvduke) [reply test_runner/src/main/kotlin/ftl/client/google/GcStorage.kt:121]: I totally forgot that the integration tests require a GCP credential. I'm actually seeing a failure on two tests including the `MultipleApksIT` test.

I think what's happening is the string that gets passed into the pattern you introduced has regex reserved characters in it, and so the pattern is invalid.

```
integration.MultipleApksIT > flank full option run FAILED
    java.util.regex.PatternSyntaxException: Illegal character range near index 23
    Uploading( file)? [app-multiple-success-debug-androidTest.apk]
                           ^
        at java.base/java.util.regex.Pattern.error(Pattern.java:2027)
        at java.base/java.util.regex.Pattern.range(Pattern.java:2825)
        at java.base/java.util.regex.Pattern.clazz(Pattern.java:2713)
        at java.base/java.util.regex.Pattern.sequence(Pattern.java:2138)
        at java.base/java.util.regex.Pattern.expr(Pattern.java:2068)
        at java.base/java.util.regex.Pattern.compile(Pattern.java:1782)
        at java.base/java.util.regex.Pattern.<init>(Pattern.java:1428)
        at java.base/java.util.regex.Pattern.compile(Pattern.java:1094)
        at utils.IntergrationTestsUtilsKt.assertContainsUploads(IntergrationTestsUtils.kt:60)
        at integration.MultipleApksIT.flank full option run(MultipleApksIT.kt:45)

integration.CustomShardingIT > flank custom sharding -- android FAILED
    java.util.regex.PatternSyntaxException: Illegal character range near index 23
    Uploading( file)? [app-multiple-success-debug-androidTest.apk]
                           ^
        at java.base/java.util.regex.Pattern.error(Pattern.java:2027)
        at java.base/java.util.regex.Pattern.range(Pattern.java:2825)
        at java.base/java.util.regex.Pattern.clazz(Pattern.java:2713)
        at java.base/java.util.regex.Pattern.sequence(Pattern.java:2138)
        at java.base/java.util.regex.Pattern.expr(Pattern.java:2068)
        at java.base/java.util.regex.Pattern.compile(Pattern.java:1782)
        at java.base/java.util.regex.Pattern.<init>(Pattern.java:1428)
        at java.base/java.util.regex.Pattern.compile(Pattern.java:1094)
        at utils.IntergrationTestsUtilsKt.assertContainsUploads(IntergrationTestsUtils.kt:60)
        at integration.CustomShardingIT.flank custom sharding -- android(CustomShardingIT.kt:73)
```

**Comment by adamvduke** [comment]
@flank-it

**Comment by joshafeinberg** [comment]
@flank-it

**Review by adamvduke — APPROVED** [review]

**Comment by ArunYogeshwaran** [comment]
Thanks for this fix - Upgrading Flank version from 21.08.0 to 23.0.10 fixed the issue for us 🚀

**Pull request merged**